combustion-android-ble/src/main/java/inc/combustion/framework/ble/BleDevice.kt

```diff
@@ -0,0 +1,253 @@
+    adapter: BluetoothAdapter
+) {
+    companion object {
+        private const val PROBE_REMOTE_RSSI_POLL_RATE_MS = 2000L
+        private const val PROBE_IDLE_TIMEOUT_POLL_RATE_MS = 1000L
+        private const val PROBE_IDLE_TIMEOUT_MS = 15000L
+        const val DISCONNECT_TIMEOUT_MS = 500L
+
+        private const val DEV_INFO_SERVICE_UUID_STRING = "0000180A-0000-1000-8000-00805F9B34FB"
+        private val SERIAL_NUMBER_CHARACTERISTIC = characteristicOf(
+            service = DEV_INFO_SERVICE_UUID_STRING,
+            characteristic = "00002A25-0000-1000-8000-00805F9B34FB"
+        )
+        private val FW_VERSION_CHARACTERISTIC = characteristicOf(
+            service = DEV_INFO_SERVICE_UUID_STRING,
+            characteristic = "00002A26-0000-1000-8000-00805F9B34FB"
+        )
+        private val HW_REVISION_CHARACTERISTIC = characteristicOf(
+            service = DEV_INFO_SERVICE_UUID_STRING,
+            characteristic = "00002A27-0000-1000-8000-00805F9B34FB"
+        )
+    }
+
+    val jobManager = JobManager()
+    protected var peripheral: Peripheral =
+        owner.lifecycleScope.peripheral(adapter.getRemoteDevice(mac)) {
+            logging {
+                /* The following enables logging in Kable
+                // engine = SystemLogEngine
+                // level = Logging.Level.Events
+                // format = Logging.Format.Multiline
+                // data = Hex
+                 */
+            }
+        }
+
+    protected val connectionMonitor = IdleMonitor()
+    val remoteRssi = AtomicInteger(0)
+    val rssi get() = remoteRssi.get()
+    var connectionState = DeviceConnectionState.OUT_OF_RANGE
+    val isConnected = AtomicBoolean(false)
+
+    fun observeOutOfRange(timeout: Long, callback: (suspend () -> Unit)? = null) {
+        jobManager.addJob(owner.lifecycleScope.launch(Dispatchers.IO) {
+            while(isActive) {
+                delay(PROBE_IDLE_TIMEOUT_POLL_RATE_MS)
+                val isIdle = connectionMonitor.isIdle(timeout)
+
+                if(isIdle) {
+                    connectionState = when(connectionState) {
+                        DeviceConnectionState.ADVERTISING_NOT_CONNECTABLE -> DeviceConnectionState.OUT_OF_RANGE
+                        DeviceConnectionState.ADVERTISING_CONNECTABLE -> DeviceConnectionState.OUT_OF_RANGE
+                        DeviceConnectionState.DISCONNECTED -> DeviceConnectionState.OUT_OF_RANGE
+                        else -> connectionState
+                    }
+                    callback?.let {
+                        it()
+                    }
+                }
+            }
+        })
+    }
+
+    fun observeRemoteRssi(callback: (suspend (rssi: Int) -> Unit)? = null) {
+        jobManager.addJob(owner.lifecycleScope.launch(Dispatchers.IO) {
+            var exceptionCount = 0;
+            while(isActive) {
+                if(isConnected.get() && mac != SimulatedProbeManager.SIMULATED_MAC) {
```

Comment:
Can the simulated probe be handled closer to the probe device implementation?

Reply:
I think so.  I put together the following design notes under DROID-62 that I think would cover that... I'll start a thread in Slack for opinions.

![image](https://user-images.githubusercontent.com/90232532/212082834-44cc3adb-69af-4fec-8ed0-dc7416034b16.png)